core/deployment/src/main/java/io/quarkus/deployment/util/WebJarUtil.java

```diff
@@ -0,0 +1,150 @@
+package io.quarkus.deployment.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.deployment.pkg.builditem.CurateOutcomeBuildItem;
+
+/**
+ * Utility for Web resource related operations
+ */
+public class WebJarUtil {
+
+    private final static String tmpDir = System.getProperty("java.io.tmpdir");
+    private static final String OWN_MEDIA_FOLDER = "META-INF/resources/";
+    private static final List<String> IGNORE_LIST = Arrays.asList("logo.png", "favicon.ico");
+
+    private WebJarUtil() {
+    }
+
+    public static Path devOrTest(AppArtifact artifact, String rootFolderInJar)
+            throws IOException {
+
+        Path path = createDir(artifact.getGroupId(), artifact.getArtifactId(), artifact.getVersion());
+        if (isEmpty(path)) {
+            ClassLoader classLoader = WebJarUtil.class.getClassLoader();
+            for (Path p : artifact.getPaths()) {
+                File artifactFile = p.toFile();
+                try (JarFile jarFile = new JarFile(artifactFile)) {
+                    Enumeration<JarEntry> entries = jarFile.entries();
+                    while (entries.hasMoreElements()) {
+                        JarEntry entry = entries.nextElement();
+                        if (entry.getName().startsWith(rootFolderInJar) && !entry.isDirectory()) {
+                            try (InputStream inputStream = jarFile.getInputStream(entry)) {
+                                String filename = entry.getName().replace(rootFolderInJar, "");
```

Comment:
To reduce the overhead for "concurrent access" case, I would pull this line up (out of the `try (InputStream inputStream = ...`) and call `continue` in case the file already exists.

Reply:
Feels like premature optimization.... let's wait and see if it's so bad that people complain ? w.d.y.t ?

Reply:
Yeah, sure.